pkg/services/object/util/chain.go

```diff
@@ -34,45 +38,178 @@ func IterateAllSplitLeaves(r HeadReceiver, addr oid.Address, h func(*object.Obje
+	info, err := r.Head(addr)
+	if err != nil {
+		return fmt.Errorf("receiving information about the object: %w", err)
+	}
+
+	switch res := info.(type) {
+	default:
+		panic(fmt.Sprintf("unexpected result of %T: %T", r, info))
+	case *object.Object:
+		h(res)
+	case *object.SplitInfo:
+		if res.SplitID() == nil {
+			return iterateV2Split(r, res, addr.Container(), h)
+		} else {
+			return iterateV1Split(r, res, addr.Container(), h)
 		}
+	}
 
-		return h(member)
-	}); err != nil {
-		return err
+	return nil
+}
+
+func iterateV1Split(r SplitInformer, info *object.SplitInfo, cID cid.ID, handler func(*object.Object) bool) error {
+	var addr oid.Address
+	addr.SetContainer(cID)
+
+	linkID, ok := info.Link()
+	if ok {
+		addr.SetObject(linkID)
+
+		linkObj, err := headFromInformer(r, addr)
+		if err != nil {
+			return err
+		}
+
+		for _, child := range linkObj.Children() {
+			addr.SetObject(child)
+
+			childHeader, err := headFromInformer(r, addr)
+			if err != nil {
+				return err
+			}
+
+			if stop := handler(childHeader); stop {
+				return nil
+			}
+		}
+
+		handler(linkObj)
+
+		return nil
+	}
+
+	lastID, ok := info.LastPart()
+	if ok {
+		addr.SetObject(lastID)
+		return iterateFromLastObject(r, addr, handler)
 	}
 
-	for i := len(leaves) - 1; i >= 0; i-- {
-		if h(leaves[i]) {
+	return errors.New("nor link, nor last object ID is found")
+}
+
+func iterateV2Split(r SplitInformer, info *object.SplitInfo, cID cid.ID, handler func(*object.Object) bool) error {
+	var addr oid.Address
+	addr.SetContainer(cID)
+
+	linkID, ok := info.Link()
+	if ok {
+		addr.SetObject(linkID)
+
+		linkObjRaw, err := r.Get(addr)
+		if err != nil {
+			return fmt.Errorf("receiving link object %s: %w", addr, err)
+		}
+
+		if stop := handler(&linkObjRaw); stop {
+			return nil
+		}
+
+		var linkObj object.Link
+		err = linkObjRaw.ReadLink(&linkObj)
+		if err != nil {
+			return fmt.Errorf("decoding link object: %w", err)
+		}
+
+		for _, child := range linkObj.Objects() {
+			addr.SetObject(child.ObjectID())
+
+			childObj, err := headFromInformer(r, addr)
+			if err != nil {
+				return fmt.Errorf("fetching child object: %w", err)
+			}
+
+			if stop := handler(childObj); stop {
+				return nil
+			}
+		}
+
+		return nil
+	}
+
+	lastID, ok := info.LastPart()
+	if ok {
+		addr.SetObject(lastID)
+		return iterateFromLastObject(r, addr, handler)
+	}
+
+	return errors.New("nor link, nor last object ID is found")
+}
+
+func iterateFromLastObject(r SplitInformer, lastAddr oid.Address, handler func(*object.Object) bool) error {
+	var idBuff []oid.ID
+	addr := lastAddr
+
+	for {
+		obj, err := headFromInformer(r, addr)
+		if err != nil {
+			return err
+		}
+
+		oID, _ := obj.ID()
+		idBuff = append(idBuff, oID)
+
+		prevOID, set := obj.PreviousID()
+		if !set {
 			break
 		}
+
+		addr.SetObject(prevOID)
+	}
+
+	for i := len(idBuff) - 1; i >= 0; i-- {
+		addr.SetObject(idBuff[i])
+
+		childObj, err := headFromInformer(r, addr)
```

Comment:
agree, looks ugly. but what if we have 1M 16K children as @roman-khimov loves? saving 1M 32-bytes IDs would be better than storing so many headers. not sure what is worse: duplicating RPCs or having 16GB memory for simple SG object put

also, this case is not expected since link object are always preferable

also, we are not expecting to have ID in the right order in SG as i remember

Reply:
> also, we are not expecting to have ID in the right order in SG as i remember

woops, no: https://github.com/nspcc-dev/neofs-api/blob/5c3535423c564fe63991812cb84d8334db1c553a/storagegroup/types.proto#L32